### ISSUE TITLE
Spin of death when writing in half-closed ssl socket

### ISSUE THREAD

**Issue by vyzo** [gambit/gambit#281]
We have a reproducible but not yet fully isolated pathological spin condition with SSL sockets (I am helping @fare debug this).

Specifically, we see spinning behaviour when forcing output to an SSL socket whose input side has been closed, either abortively with an error or with a graceful protocol-level close (it's with websockets fwiw). 
The backtrace with gdb indicates that it's stuck in a `___device_tcp_client_write_raw_virt` loop, getting ssl errors.
After multiple interrupts (on two different stuck programs exhibiting the same behaviour) we have this common stack trace:
```
#1  0x00007f697d6aa0e7 in ERR_get_state ()
   from /nix/store/1w1qqziw1vngg075slbjmg72wyz5cx28-openssl-1.0.2l/lib/libcrypto.so.1.0.0
#2  0x00007f697d6aa3f6 in get_error_values ()
   from /nix/store/1w1qqziw1vngg075slbjmg72wyz5cx28-openssl-1.0.2l/lib/libcrypto.so.1.0.0
#3  0x00007f697e584235 in ___device_tcp_client_write_raw_virt ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#4  0x00007f697e5875ef in ___os_device_stream_write ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#5  0x00007f697e598023 in ___H__20___kernel ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#6  0x00007f697e56df66 in ___trampoline ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#7  0x00007f697e56e0f7 in ___call ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#8  0x00007f697e56e1f3 in ___run ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#9  0x00007f697e56f256 in ___setup ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#10 0x00007f697e56b9d0 in ___main ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#11 0x00007f697e57ab30 in ___main_char ()
   from /nix/store/ngaibjpm6j3zcwi8n08ikk7n0l70lknw-gambit-4.8.8-d5e06fc/lib/libgambit.so
#12 0x00007f697d214530 in __libc_start_main ()
   from /nix/store/vzhrb2vwp5z5ijzaigrah9pifp3nn41p-glibc-2.25/lib/libc.so.6
#13 0x0000000000401c0a in _start () at ../sysdeps/x86_64/start.S:120
```

I will try to reproduce this with debug enabled so that we can get a cleaner picture of the situation.

**Comment by vyzo** [comment]
Upon some investigation in the code and the exhibited behaviour, I believe the threads get stuck in the `clear_tls_error_queue` loop.

The code reads:
```
___HIDDEN void clear_tls_error_queue
   ___PVOID
{
  while (ERR_peek_error() != 0)
    {
      /* ERR_print_errors_fp (stderr); */
    }
  ERR_clear_error();
}

```

The manpage reads:
```
       ERR_get_error() returns the earliest error code from the thread's error queue and
       removes the entry. This function can be called repeatedly until there are no more error
       codes to return.

       ERR_peek_error() returns the earliest error code from the thread's error queue without
       modifying it.

```

I believe this is a bug and this function spins when there is an error. Most likely you wanted to use `ERR_get_error` in the loop, and not the non-modifying `ERR_peek_error`.

**Comment by vyzo** [comment]
I am pretty sure your code intended to get_error or clear in the commented out part; now it's just a spin. 

The patch removes it altogether as `ERR_clear_error` does the job. From the manpage:
```
       ERR_clear_error() empties the current thread's error queue.

```

**Comment by adam-7** [comment]
The real, scalable solution involves kicking out the SSL logic from Gambit altogether and implementing userland ports.

Right now the SSL is really a patch on the TCP code, which is quite crazy as OpenSSL is a buggy, broken library with some of the awkwardest specifics in the open source world. It's a pure coincidence that this implementation style works, and it could easily break.

Furthermore having the SSL logic in Gambit makes it more difficult to refine degradation behavior in the failure cases, and SSL for sure comes with very many of these.

If you want a userland SSL channel implementation let me know, I think also it's on the wiki/ML, many weeks went into it.


So to sum up I'd say given how broken OpenSSL is, it's only a coincidence that the particular channel implementation you are using, works. It's in Gambit's interest (as a sleek system) to externalize the SSL logic ASAP.
